compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs

```diff
@@ -794,23 +805,192 @@ generateSrcPkgFromLf config pkg = do
         , "{-# OPTIONS_GHC -Wno-unused-imports -Wno-missing-methods #-}"
         ]
 
--- | Returns 'True' if an LF type contains a reference to an
--- old-style typeclass. See 'tconIsOldTypeclass' for more details.
-typeHasOldTypeclass :: Env -> LF.Type -> Bool
-typeHasOldTypeclass env = \case
-    LF.TVar _ -> False
-    LF.TCon tcon -> tconIsOldTypeclass env tcon
-    LF.TSynApp _ _ -> False
-        -- Type synonyms came after the switch to new-style
-        -- typeclasses, so we can assume there are no old-style
-        -- typeclasses being referenced.
-    LF.TApp a b -> typeHasOldTypeclass env a || typeHasOldTypeclass env b
-    LF.TBuiltin _ -> False
-    LF.TForall _ b -> typeHasOldTypeclass env b
-    LF.TStruct fields -> any (typeHasOldTypeclass env . snd) fields
-    LF.TNat _ -> False
-
--- | Determine whether a typecon refers to an old-style
+-- | A reference that can appear in a type or expression. We need to track
+-- all of these in order to control what gets hidden in data-dependencies.
+data Ref
+    = RTypeCon (LF.Qualified LF.TypeConName)
+        -- ^ necessary to track hidden types and old-style typeclasses
+    | RTypeSyn (LF.Qualified LF.TypeSynName)
+        -- ^ necessary to track hidden new-style typeclasses
+    | RValue (LF.Qualified LF.ExprValName)
+        -- ^ necessary to track hidden typeclass instances
+    deriving (Eq, Ord, Show, Generic)
+
+instance Hashable Ref
+
+refPackage :: Ref -> LF.PackageRef
+refPackage = \case
+    RTypeCon q -> LF.qualPackage q
+    RTypeSyn q -> LF.qualPackage q
+    RValue q -> LF.qualPackage q
+
+refModule :: Ref -> LF.ModuleName
+refModule = \case
+    RTypeCon q -> LF.qualModule q
+    RTypeSyn q -> LF.qualModule q
+    RValue q -> LF.qualModule q
+
+type RefGraph = HMS.HashMap Ref (Bool, [Ref])
+
+-- | Calculate the set of all references that should be hidden.
+buildHiddenRefMap :: Env -> HMS.HashMap Ref Bool
+buildHiddenRefMap env =
+    case leastFixedPointBy (||) refGraphList of
+        Left ref -> error
+            ("Internal error: missing reference in RefGraph "
+            <> show ref)
+        Right m -> m
+  where
+    refGraphList :: [(Ref, Bool, [Ref])]
+    refGraphList =
+        [ (ref, hidden, deps)
+        | (ref, (hidden, deps)) <- HMS.toList refGraph
+        ]
+
+    refGraph :: RefGraph
+    refGraph = foldl visitRef HMS.empty (rootRefs env)
+
+    visitRef :: RefGraph -> Ref -> RefGraph
+    visitRef !refGraph ref
+        | HMS.member ref refGraph
+            = refGraph -- already in the map
+        | ref == RTypeCon erasedTCon
+            = HMS.insert ref (True, []) refGraph -- Erased is always erased
+        | refModule ref == LF.ModuleName ["DA", "Generics"]
+            = HMS.insert ref (True, []) refGraph -- DA.Generics is not supported
+        | LF.PRImport pkgId <- refPackage ref
+        , Set.member pkgId (configDependencyPackages (envConfig env))
```

Comment:
Will we ever look up a reference to a `dependency`? Presumably the case where this matters is for reexported typeclasses. Do we check if the original typeclass is hidden or if the new typeclass is hidden? I think I don’t really understand why we need to “white-list” this so maybe it would be good to expand on this in a comment.

Reply:
I didn't check to see if it made a significant difference, but this whitelist is for the case where the data-dependency shares a dependency with the package being built. It doesn't check re-exports, it's just a check for actual references to the same dependencies.

The only case where this is logically relevant is if both the package being built & the data-dependencies depend on the same package, and that package contains something that would otherwise be erased. That's pretty rare. OTOH, this check prevents us from building up the a complicated dependency graph for daml-stdlib / daml-prim modules each time, on the assumption that the SDK hasn't changed.

Given that the case where it matters is quite rare, I can try removing this and see if it makes a difference to compile times. If removing it makes it significantly worse, I'll keep the whitelist and add a comment, and then see if we can just eat the cost once we start building this map once per package instead of once per module.

Reply:
Thanks for the explanation! Happy to keep it for this PR if we add a comment.